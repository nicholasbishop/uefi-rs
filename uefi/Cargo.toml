[package]
name = "uefi"
version = "0.26.0"
readme = "README.md"
description = "Safe and easy-to-use wrapper for building UEFI apps."

authors.workspace = true
categories.workspace = true
edition.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[features]
default = ["panic-on-logger-errors"]
alloc = []
global_allocator = []
logger = []
# Ignore text output errors in logger as a workaround for firmware issues that
# were observed on the VirtualBox UEFI implementation (see uefi-rs#121).
# In those cases, this feature can be excluded by removing the default features.
panic-on-logger-errors = []
# Generic gate to code that uses unstable features of Rust. You usually need a nightly toolchain.
unstable = []

[dependencies]
bitflags.workspace = true
log.workspace = true
ptr_meta.workspace = true
ucs2 = "0.3.2"
uefi-macros = "0.13.0"
uefi-raw = "0.5.0"
uguid.workspace = true

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]
