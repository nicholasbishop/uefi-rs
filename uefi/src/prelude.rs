//! This module is used to simplify importing the most common UEFI types.
//!
//! This includes the system table types, `Status` codes, etc.

pub use crate::{Handle, ResultExt, Status};

// Import the basic table types.
pub use crate::table::boot::BootServices;
pub use crate::table::runtime::RuntimeServices;
pub use crate::table::{Boot, SystemTable};

// Import the macro for creating the custom entry point.
pub use uefi_macros::entry;
