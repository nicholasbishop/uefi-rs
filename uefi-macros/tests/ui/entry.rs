#![no_main]
#![feature(abi_efiapi)]

use uefi::prelude::*;
use uefi_macros::entry;

mod good_entry {
    use super::*;

    #[entry]
    fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
        Status::SUCCESS
    }
}

mod bad_attr_arg {
    use super::*;

    #[entry(some_arg)]
    extern "C" fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
        Status::SUCCESS
    }
}

mod bad_abi_modifier {
    use super::*;

    #[entry]
    extern "C" fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
        Status::SUCCESS
    }
}

mod bad_async {
    use super::*;

    #[entry]
    async fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
        Status::SUCCESS
    }
}

mod bad_const {
    use super::*;

    #[entry]
    const fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
        Status::SUCCESS
    }
}

mod bad_generic {
    use super::*;

    #[entry]
    fn efi_main<T>(_handle: Handle, _st: SystemTable<Boot>) -> Status {
        Status::SUCCESS
    }
}

mod bad_args {
    use super::*;

    #[entry]
    fn efi_main(_handle: Handle, _st: SystemTable<Boot>, _x: usize) -> bool {
        false
    }
}

mod bad_return_type {
    use super::*;

    #[entry]
    fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> bool {
        false
    }
}
