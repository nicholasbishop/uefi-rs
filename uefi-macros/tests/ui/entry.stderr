error: Entry attribute accepts no arguments
  --> $DIR/entry.rs:19:13
   |
19 |     #[entry(some_arg)]
   |             ^^^^^^^^

error: Entry method must have no ABI modifier
  --> $DIR/entry.rs:20:5
   |
20 |     extern "C" fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
   |     ^^^^^^^^^^

error: Entry method must have no ABI modifier
  --> $DIR/entry.rs:29:5
   |
29 |     extern "C" fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
   |     ^^^^^^^^^^

error: Entry method should not be async
  --> $DIR/entry.rs:38:5
   |
38 |     async fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
   |     ^^^^^

error: Entry method should not be const
  --> $DIR/entry.rs:47:5
   |
47 |     const fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> Status {
   |     ^^^^^

error: Entry method should not be generic
  --> $DIR/entry.rs:56:17
   |
56 |     fn efi_main<T>(_handle: Handle, _st: SystemTable<Boot>) -> Status {
   |                 ^

error[E0308]: mismatched types
  --> $DIR/entry.rs:65:8
   |
65 |     fn efi_main(_handle: Handle, _st: SystemTable<Boot>, _x: usize) -> bool {
   |        ^^^^^^^^ incorrect number of function parameters
   |
   = note: expected fn pointer `extern "efiapi" fn(uefi::Handle, uefi::table::SystemTable<uefi::table::Boot>) -> uefi::Status`
                 found fn item `extern "efiapi" fn(uefi::Handle, uefi::table::SystemTable<uefi::table::Boot>, usize) -> bool {bad_args::efi_main}`

error[E0308]: mismatched types
  --> $DIR/entry.rs:74:8
   |
74 |     fn efi_main(_handle: Handle, _st: SystemTable<Boot>) -> bool {
   |        ^^^^^^^^ expected struct `uefi::Status`, found `bool`
   |
   = note: expected fn pointer `extern "efiapi" fn(uefi::Handle, uefi::table::SystemTable<_>) -> uefi::Status`
                 found fn item `extern "efiapi" fn(uefi::Handle, uefi::table::SystemTable<_>) -> bool {bad_return_type::efi_main}`
